cmd/kubectl-gadget/top/top.go

```diff
@@ -47,11 +59,153 @@ func addCommonTopFlags(
+type TopGadget[Stats any] struct {
+	sync.Mutex
+
+	name           string
+	commonTopFlags *CommonTopFlags
+	params         map[string]string
+	parser         TopParser[Stats]
+	nodeStats      map[string][]*Stats
+	colMap         columns.ColumnMap[Stats]
+}
+
+func (g *TopGadget[Stats]) Run(args []string) error {
+	var err error
+
+	if len(args) == 1 {
+		g.commonTopFlags.OutputInterval, err = strconv.Atoi(args[0])
+		if err != nil {
+			return commonutils.WrapInErrInvalidArg("<interval>",
+				fmt.Errorf("%q is not a valid value", args[0]))
+		}
+	} else {
+		g.commonTopFlags.OutputInterval = top.IntervalDefault
+	}
+
+	sortByColumns := strings.Split(g.commonTopFlags.SortBy, ",")
+	cols := columns.MustCreateColumns[Stats]()
+	_, invalidCols := cols.VerifyColumnNames(sortByColumns)
+
+	if len(invalidCols) > 0 {
+		return commonutils.WrapInErrInvalidArg("--sort", fmt.Errorf("\"%v\" is/are not a recognized column(s) to sort by", strings.Join(invalidCols, ",")))
+	}
+	g.commonTopFlags.ParsedSortBy = sortByColumns
+
+	if g.params == nil {
+		g.params = make(map[string]string)
+	}
+	g.params[top.MaxRowsParam] = strconv.Itoa(g.commonTopFlags.MaxRows)
+	g.params[top.IntervalParam] = strconv.Itoa(g.commonTopFlags.OutputInterval)
+	g.params[top.SortByParam] = g.commonTopFlags.SortBy
+
+	config := &utils.TraceConfig{
+		GadgetName:       g.name,
+		Operation:        gadgetv1alpha1.OperationStart,
+		TraceOutputMode:  gadgetv1alpha1.TraceOutputModeStream,
+		TraceOutputState: gadgetv1alpha1.TraceStateStarted,
+		CommonFlags:      &g.commonTopFlags.commonFlags,
+		Parameters:       g.params,
+	}
+
+	// when params.Timeout == interval it means the user
+	// only wants to run for a given amount of time and print
+	// that result.
+	singleShot := g.commonTopFlags.commonFlags.Timeout == g.commonTopFlags.OutputInterval
+
+	// start print loop if this is not a "single shot" operation
+	if singleShot {
+		g.PrintHeader()
+	} else {
+		g.StartPrintLoop()
+	}
+
+	if err := utils.RunTraceStreamCallback(config, g.Callback); err != nil {
+		return commonutils.WrapInErrRunGadget(err)
+	}
+
+	if singleShot {
+		g.PrintStats()
+	}
+
+	return nil
+}
+
+func (g *TopGadget[Stats]) StartPrintLoop() {
+	go func() {
+		ticker := time.NewTicker(time.Duration(g.commonTopFlags.OutputInterval) * time.Second)
+		g.PrintHeader()
+		for {
+			<-ticker.C
+			g.PrintHeader()
+			g.PrintStats()
+		}
+	}()
+}
+
+func (g *TopGadget[Stats]) PrintHeader() {
+	if g.commonTopFlags.commonFlags.OutputMode == commonutils.OutputModeJSON {
+		return
+	}
+
+	if term.IsTerminal(int(os.Stdout.Fd())) {
+		utils.ClearScreen()
+	} else {
+		fmt.Println("")
+	}
+
+	fmt.Println(g.parser.BuildColumnsHeader())
+}
+
+func (g *TopGadget[Stats]) Callback(line string, node string) {
+	g.Lock()
+	defer g.Unlock()
+
+	stats, err := g.parser.UnmarshalStats(line, node)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s", err.Error())
+		return
+	}
+
+	g.nodeStats[node] = stats
+}
```

Comment:
This could be replaced by a generic Events type like so:
```suggestion
type Event[T any] struct {
	Error string `json:"error,omitempty"`
	Stats []*T   `json:"stats,omitempty"`
}

func (g *TopGadget[Stats]) Callback(line string, node string) {
	var ev Event[Stats]

	err := json.Unmarshal([]byte(line), &ev)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s", commonutils.WrapInErrUnmarshalOutput(err, line))
		return
	}
	if ev.Error != "" {
		fmt.Fprintf(os.Stderr, "Error: failed on node %q: %s", node, ev.Error)
	}

	g.Lock()
	defer g.Unlock()

	g.nodeStats[node] = ev.Stats
}
```

Reply:
Did not thought about making the `Event` generic.

Would it be even better then if i generalize all `Event` structs of the `top` gadgets and put one definition of the Event struct in `pkg/gadgets/top/top.go`? Maybe even calling it `TopEvent`?

Then it would be only defined once and we could use it in this file as well

Reply:
Yeah, I would remove the specific ones and replace them with this generic one, too.